### ISSUE TITLE
Comments and States

### ISSUE THREAD

**Issue by tehsphinx** [exercism/automated-analysis#38]
Sorry if this is kind of late... Here a few points I think the current interface definition still lacks.

1) I think we should never "let the student go" without a comment. If there is nothing to improve on the solution we can still give them valuable advice e.g. in the form of a pointer to further reading. This is the undisputed No 1 positive feedback we get from students on the Go track where we do this since a few months. (introduced by @bitfield).
This does not have to be done in the analyzer though. We could add that in the website and pull the `random tip` from a repository source where the tips can be added/edited. If it is done on the `analyzer` (which I think would not ideal) we don't need the status "approve". There never would be approval without a comment.

   In my mind the student gets this `tip` when they get the approval. That way every student gets one tip per exercise.

2) I assume the comments are output in the `order` the analyzer adds them to the output file so we should have the ability to `sort` the comments.
But I am missing a way to `categorize` the comments. I propose to have at least 3 or 4 categories:
- `To Do`: This is for comments that block automatic approval.
- `Improvement`: This is for comments that make the code better, more readable or more idiomatic but would not block an approval.
- `Thoughts`: Further thoughts on the solution like alternatives ways to solve this or hints on _exercise related_ materials for further study.
- `Blocks`: This category is defined by it's type. We called it a `block comment`. It is typically a short introduction to a topic and links to further material. For `Go` we have blocks e.g. about `benchmarking`, `regex`, `pprof: allocations`. This category was always shown last and referenced to in other comments. e.g. `For further information see section about "benchmarking" below.`

  I suggest the categories are printed to the student in the order they are mentioned here. Most important (blocking) first, etc.

  Also I would like the student to know what is what. This helps avoiding confusion for example about a `thought` being something the student has to do to get approved. In exalysis we declared this very clearly with a sentence before each category:
`Could you have a look at the following points so I can approve the solution?` (To Do)
`Here are some thoughts for further improvement:` (Improvements)
`Here are some further thoughts on this exercise:` (Thoughts)
`Here's something you might find interesting. It's not necessarily related to this exercise, but it will help you develop your general Go knowledge:` (Random Tip. See 1))

   Note: I realize this seems strange if all your analyzer did for now is return one comment.

**Comment by tehsphinx** [comment]
@iHiD let me know if you want me to separate this into 2 issues

**Comment by iHiD** [comment]
@tehsphinx This is all great. Thank you! We'll use it in our internal product-dev conversations about how we display comments to students, which is something that needs a lot of thought/work.

It's worth me reiterating that for the first cut of all this, the only state we will honour is `approve_as_optimal`. We will not be returning messages until we have done a wad of product-dev on that. So for the first cut, please do make use of that, even if in the longer run you want to add tips etc. In terms of where those tips live, that's another big discussion we need to have.

So, in summary, this discussion is one we will have, and is super-valuable, but for our first step forward, we're all about approving optimal solutions to keep people moving through the tracks.

**Comment by tehsphinx** [comment]
Looking forward to all those discussions :)
